lib/include/tree_sitter/api.h

```diff
@@ -602,6 +635,147 @@ int64_t ts_tree_cursor_goto_first_child_for_byte(TSTreeCursor *, uint32_t);
 
 TSTreeCursor ts_tree_cursor_copy(const TSTreeCursor *);
 
+/*******************/
+/* Section - Query */
+/*******************/
+
+/**
+ * Create a new query from a string containing one or more S-expression
+ * patterns. The query is associated with a particular language, and can
+ * only be run on syntax nodes parsed with that language.
+ *
+ * If all of the given patterns are valid, this returns a `TSQuery`.
+ * If a pattern is invalid, this returns `NULL`, and provides two pieces
+ * of information about the problem:
+ * 1. The byte offset of the error is written to the `error_offset` parameter.
+ * 2. The type of error is written to the `error_type` parameter.
+ */
+TSQuery *ts_query_new(
+  const TSLanguage *language,
+  const char *source,
+  uint32_t source_len,
+  uint32_t *error_offset,
+  TSQueryError *error_type
+);
+
+/**
+ * Delete a query, freeing all of the memory that it used.
+ */
+void ts_query_delete(TSQuery *);
+
+/**
+ * Get the number of patterns in the query.
+ */
+uint32_t ts_query_pattern_count(const TSQuery *);
+
+/**
+ * Get the predicates for the given pattern in the query.
+ */
+const TSQueryPredicateStep *ts_query_predicates_for_pattern(
+  const TSQuery *self,
+  uint32_t pattern_index,
+  uint32_t *length
+);
+
+/**
+ * Get the number of distinct capture names in the query, or the number of
+ * distinct string literals in the query.
+ */
+uint32_t ts_query_capture_count(const TSQuery *);
+uint32_t ts_query_string_count(const TSQuery *);
+
+/**
+ * Get the name and length of one of the query's captures, or one of the
+ * query's string literals. Each capture and string is associated with a
+ * numeric id based on the order that it appeared in the query's source.
+ */
+const char *ts_query_capture_name_for_id(
+  const TSQuery *,
+  uint32_t id,
+  uint32_t *length
+);
+const char *ts_query_string_value_for_id(
```

Comment:
Why not have capture names and this kind of general "strings" (only used by predicates, and not the query engine itself?) share a single "namespace" of interned string id:s, and a single function to look them up? It could still have all the capture names in the beginning, if there is a need to index a dense array of captures or something. It would simplify FFI by only managing one table of interned strings, and not two.

Reply:
The main reason I lead toward keeping them separate is that callers will need to use them slightly differently. When creating bindings to this API, the `capture_names` need to stored in an array that will be used every time you run a query, whereas the `strings` only need to be loaded temporarily, while constructing a query.

With this approach, we can preallocate and populate the `captureNames` array using `ts_query_capture_count / ts_query_capture_name_for_id`, and preallocate and populate a separate `stringValues` array using `ts_query_string_count / ts_query_string_value_for_id`. Then, at the end of the query constructor function, we can simply drop the `stringValues` array.

> It could still have all the capture names in the beginning, if there is a need to index a dense array of captures or something.

That's true. With that approach, we could at least *shrink* the array once we are done using the `strings`. But in order to do this, callers still need to know how many captures there are, so there still needs to be a separate `ts_query_capture_count` function.

In the end, I think the APIs are slightly more obvious in their meaning if `captures` and `strings` are just presented as two separate arrays, where each one has identical APIs for accessing its length and its contents.

Reply:
I see  what you mean, though this seems to assume very specific details how you would use the API. In the first iteration I would do as much as possible lua-side, at least the non-hot preprocessing of the queries. So this would just imply needing two code paths (or helper functions) for converting an internal id to a lua string instead of just one. Though it is not much duplicated code, probably only like 3-5 lines.

Reply:
Yeah, good point. I'm writing the API with the assumption that FFI has some noticeable cost, based on experience with JavaScript via Node and WebAssembly. If you aren't keeping a managed copy of the captures array, then none of the things that I said above are applicable.

I think I'll keep it separate for now, though I see your point about the potential reduction in API functions.

On a related note, in 1af85dc3f758e5b4e2179fec81f95b89dadc26e3 I did delete two APIs related to captures and string values, which I realized were unused by either of my bindings. So now there are slightly fewer 😁.